server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3801,6 +3833,32 @@ private ActionListener<Releasable> wrapPrimaryOperationPermitListener(final Acti
         });
     }
 
+    /**
+     * Immediately delays operations and uses the {@code executor} to wait for in-flight operations to finish and then acquires all
+     * permits. When all permits are acquired, the provided {@link ActionListener} is called under the guarantee that no new operations are
+     * started. Delayed operations are run once the {@link Releasable} is released or if a failure occurs while acquiring all permits; in
+     * this case the {@code onFailure} handler will be invoked after delayed operations are released.
+     *
+     * @param onAcquired {@link ActionListener} that is invoked once acquisition is successful or failed. This listener should not throw.
+     * @param timeout    the maximum time to wait for the in-flight operations block
+     * @param timeUnit   the time unit of the {@code timeout} argument
+     * @param executor   executor on which to wait for in-flight operations to finish and acquire all permits
+     */
+    public void blockOperations(
+        final ActionListener<Releasable> onAcquired,
+        final long timeout,
+        final TimeUnit timeUnit,
+        final Executor executor
+    ) {
+        indexShardOperationPermits.delayOperations();
+        // In case indexing is paused on the shard, suspend throttling so that any currently paused task can
+        // go ahead and release the indexing permit it holds.
+        suspendThrottling();
+        indexShardOperationPermits.waitUntilBlocked(ActionListener.assertOnce(onAcquired), timeout, timeUnit, executor);
+        // TODO: Does this do anything ? Looks like the relocated shard does not have throttling enabled
+        resumeThrottling();
```

Comment:
I think we can just use `blockOperations` instead, rather than expose the underlying private methods.

Also, we should resume the throttle in the action listener instead.

```suggestion
        // In case indexing is paused on the shard, suspend throttling so that any currently paused task can
        // go ahead and release the indexing permit it holds.
        suspendThrottling();
        try {
        indexShardOperationPermits.blockOperations(ActionListener.runAfter(onAcquired, this::resumeThrottling), timeout, timeUnit, executor);
        } catch (IndexShardClosedException e) {
            resumeThrottling();
            throw e;
        }
```

We can change the exception throw in `delayOperations` to use the listener instead in a follow-up (I worry it *might* have side-effects, so best to tackle separately). Happy to leave with the catch too though not too pretty.

Reply:
Yes, or rather I do not want to expose the difference from the permits class.